webdriver/tests/support/asserts.py

```diff
@@ -86,14 +85,16 @@ def assert_response_headers(headers):
     # In Python 2, HTTPResponse normalizes header keys to lowercase, whereas
     # Python 3 preserves the case. See
     # https://github.com/web-platform-tests/wpt/pull/22858#issuecomment-612656097
-    if PY3:
+    # For Firefox, the response header names are converted to lowercase strings for consistent lookups.
+    # https://www.selenium.dev/selenium/docs/api/javascript/module/selenium-webdriver/lib/http_exports_Response.html
+    try:
         assert 'Content-Type' in headers
```

Comment:
So I understand properly; this has always been an 'issue' (spec-compliant issue, perhaps),  with geckodriver. Back in py2, we didn't notice because HTTPResponse was normalizing the cases. In py3, HTTPResponse no longer normalizes the case so this fails for geckodriver?

I wonder if we should be normalizing headers then... cc @jgraham @Hexcles

Reply:
Sorry I missed this. Python 2 seems to normalize (lower) the keys when parsing and storing the HTTP message (I haven't checked the source code, but just an observation based on https://github.com/web-platform-tests/wpt/pull/22858#issuecomment-612656097), while Python 3 in fact does case-insensitive lookup while preserving the casing, which arguably is a better approach:

https://github.com/python/cpython/blob/7bcc6456ad4704da9b287c8045768fa53961adc5/Lib/email/message.py#L462-L468
https://github.com/python/cpython/blob/7bcc6456ad4704da9b287c8045768fa53961adc5/Lib/email/message.py#L423-L424

It was really puzzling to me why #22858 was needed in the first place.

Turns out, we actually coerced `headers` into a dict...
https://github.com/web-platform-tests/wpt/blob/654d5f8ffb8e619b04902c1d2a5ff5b6cf7950ea/tools/webdriver/webdriver/transport.py#L43

We might want to stop doing that, or at least use this class from `wptserve` instead
https://github.com/web-platform-tests/wpt/blob/42a4b17fb42a33b5eef5c8430f4e64a19e0c235e/tools/wptserve/wptserve/request.py#L373

Reply:
Original discussion appears to be here: https://github.com/web-platform-tests/wpt/pull/14074

Reply:
Robert and I chatted about this, and we think making `Response.from_http` use a class like `RequestHeaders` would be best, as it both solves the case sensitivity problem as well as supports multiple headers (which the current code doesn't do).

The downside is that it would introduce a dependency from webdriver to wptserve, or otherwise require code duplication. At this point I am leaning towards code duplication, though it would be nice to perhaps introduce a WPT 'utils' library that everything can depend on? (Generally speaking wpt is quite messy atm, with multiple 'subprojects' that are rarely standalone subprojects anymore.)

cc @jgraham and @gsnedders

Reply:
Actually, looks like `RequestHeaders` assumes a fair amount about its input headers that isn't true for the usage in transport.py, where the headers are just a list of tuples. So the init code is slightly different at least.

Reply:
Playing with this in https://github.com/web-platform-tests/wpt/pull/25680